src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,155 @@
+                    currentNode.next = newNode;
+                    break;
+                }
+                currentNode = getNextNode(currentNode);
+            }
+        }
+        size++;
     }
 
     @Override
     public V getValue(K key) {
-        return null;
+        Node<K, V> node = getNodeByKey(key);
+        return node != null ? node.value : null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    public boolean containsKey(K key) {
+        return getNodeByKey(key) != null;
+    }
+
+    public boolean containsValue(V value) {
+        for (Node<K, V> node : storage) {
+            while (node != null) {
+                if (Objects.equals(node.value, value)) {
+                    return true;
+                }
+                node = node.next;
+            }
+        }
+        return false;
+    }
+
+    public K remove(K key) {
+        Node<K, V> node = getNodeByKey(key);
+        if (node != null) {
+            int index = Math.abs(getHash(key) % storage.length);
+            if (storage[index].key.equals(key)) {
+                storage[index] = node.next;
+            } else {
+                while (node.hasNext()) {
+                    if (node.next.key.equals(key)) {
+                        node.next = node.next.next;
+                        break;
+                    }
+                    node = node.next;
+                }
+            }
+            size--;
+            return key;
+        }
+        return null;
+    }
+
+    public void putAll(Map<K, V> m) {
+        for (Map.Entry<K, V> value : m.entrySet()) {
+            put(value.getKey(), value.getValue());
+        }
+    }
+
+    public void clear() {
+        storage = new Node[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
+    private Node<K, V> getNextNode(Node<K, V> currentNode) {
+        return currentNode.next == null ? null : currentNode.next;
+    }
+
+    private void resize() {
+        loadFactor = loadFactor * DEFAULT_SIZE_MULTIPLAYER;
+        size = DEFAULT_SIZE;
+        Node<K, V>[] oldStorage = storage;
+        storage = new Node[storage.length * DEFAULT_SIZE_MULTIPLAYER];
+        for (Node<K, V> node : oldStorage) {
+            while (node != null) {
+                put(node.key, node.value);
+                node = node.next;
+            }
+        }
+    }
+
+    private Node<K, V> getNodeByKey(K key) {
+        int index = Math.abs(getHash(key) % storage.length);
+        Node<K, V> currentNode = storage[index];
+        while (currentNode != null) {
+            if (Objects.equals(currentNode.key, key)) {
+                return currentNode;
+            }
+            currentNode = getNextNode(currentNode);
+        }
+        return null;
+    }
+
+    private int getHash(K key) {
+        return key == null ? 0 : key.hashCode();
+    }
+
+    private static class Node<K, V> {
```

Comment:
As I know we need to have int hash in Node too. But I don't sure, so it will be better to change it, after mentor review, if it will be needed.

Reply:
It is needed if you know that the hash will not change.  But there was no such task, it will complicate the logic if you need to search for an object with a changed hash.